STL_Extension/include/CGAL/for_each.h

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2020 GeometryFactory (France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: LGPL-3.0-or-later OR LicenseRef-Commercial
+//
+// Author(s) : Simon Giraudot
+
+#ifndef CGAL_FOR_EACH_H
+#define CGAL_FOR_EACH_H
+
+#include <CGAL/iterator.h>
+
+#ifdef CGAL_LINKED_WITH_TBB
+#include <tbb/parallel_for.h>
+#include <tbb/blocked_range.h>
+#include <tbb/scalable_allocator.h>
+#endif // CGAL_LINKED_WITH_TBB
+
+/*
+  CGAL::for_each<ConcurrencyTag>(Range, Function) does the following:
+
+  - if Sequential_tag is used, apply Function to all elements of Range
+  - if Parallel_tag is used:
+    * static assert that TBB is available
+    * if Range has random access iterators, use a TBB parallel_for
+    loop to apply Function to all elements of Range
+    * if Range doesn't have random access iterators, first copy the
+    iterators in a vector and then use a TBB parallel_for loop to
+    apply Function to all elements of Range
+*/
+
+namespace CGAL {
+
+namespace internal {
+
+// To emulate either a "break" in sequential and a "return" in
+// parallel, we use an interal throw/catch mechanism
+struct stop_for_each : public std::exception { };
+
+template <typename RangeRef, typename IteratorCategory>
+void for_each (RangeRef range,
+               const std::function<void(typename std::iterator_traits
+                                        <typename Range_iterator_type<RangeRef>::type>::reference)>& functor,
+               const Sequential_tag&,
+               IteratorCategory)
+{
+  for (typename std::iterator_traits
+         <typename Range_iterator_type<RangeRef>::type>::reference r : range)
+    try
+    {
+      functor(r);
+    }
+    catch (const stop_for_each&)
+    {
+      break;
+    }
+}
```

Comment:
The use of exceptions here seems a huge hammer for a small task. The functors could return somethings, instead of `void`, to signal need for break. It could be a simple `bool`, or better: an enum-class.

Reply:
Indeed… I remember that it somehow made sense when I first used that, but a return value is much simpler/efficient/better. I fixed it.